File: 双指针/15.三数之和.go
package 双指针

import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	n := len(nums)
	res := make([][]int, 0, n-2)
	for i := 0; i < n-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		if nums[i]+nums[i+1]+nums[i+2] > 0 {
			break
		}
		if nums[i]+nums[n-1]+nums[n-2] < 0 {
			continue
		}
		for j, k := i+1, n-1; j < k; {
			tmp := nums[i] + nums[j] + nums[k]
			if tmp > 0 {
				k--
			} else if tmp < 0 {
				j++
			} else {
				arr := []int{nums[i], nums[j], nums[k]}
				res = append(res, arr)
				for j < k && nums[j] == nums[j+1] {
					j++
				}
				for j < k && nums[k] == nums[k-1] {
					k--
				}
				j++
				k--
			}
		}
	}
	return res
}

File: 双指针/15.三数之和.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        int n=nums.length;
        for(int i=0;i<nums.length-2;i++){
            if(i>0 && nums[i-1]==nums[i]){
                continue;
            }
            if(nums[i] + nums[i+1] + nums[i+2] >0){
                break;
            }
            if(nums[i]+nums[n-1]+nums[n-2] < 0){
                continue;
            }
            for(int j=i+1,k=n-1;j<k;){
                int s=nums[i]+nums[j]+nums[k];
                if(s>0){
                    k--;
                }else if(s<0){
                    j++;
                }else{
                    List<Integer> list = new ArrayList<>();
                    list.add(nums[i]);
                    list.add(nums[j]);
                    list.add(nums[k]);
                    res.add(list);
                    while(j<k&&nums[j]==nums[j+1])j++;
                    while(j<k&&nums[k]==nums[k-1])k--;
                    j++;
                    k--;
                }
            }
        }
        return res;
    }
}